src/Compilers/Core/Portable/Text/SourceText.cs

```diff
@@ -102,21 +102,56 @@ public static SourceText From(string text, Encoding? encoding = null, SourceHash
             return new StringText(text, encoding, checksumAlgorithm: checksumAlgorithm);
         }
 
+        /// <summary>
+        /// Constructs a <see cref="SourceText"/> from text in a <see cref="StringBuilder"/>.
+        /// </summary>
+        /// <param name="stringBuilder"><see cref="StringBuilder"/> that contains the text.</param>
+        /// <param name="encoding">
+        /// Encoding of the file that the text was read from or is going to be saved to.
+        /// <see langword="null"/> if the encoding is unspecified.
+        /// If the encoding is not specified, the resulting <see cref="SourceText"/> isn't debuggable.
+        /// If an encoding-less <see cref="SourceText"/> is written to a file, <see cref="Encoding.UTF8"/> shall be used as a default.
+        /// </param>
+        /// <param name="checksumAlgorithm">
+        /// Hash algorithm to use to calculate checksum of the text that's saved to PDB.
+        /// </param>
+        /// <exception cref="ArgumentNullException"><paramref name="stringBuilder"/> is <see langword="null"/>.</exception>
+        /// <exception cref="ArgumentException"><paramref name="checksumAlgorithm"/> is not supported.</exception>
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
+        public static SourceText From(
+            StringBuilder stringBuilder,
+            Encoding? encoding = null,
+            SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1)
+#pragma warning restore RS0026 // Do not add multiple public overloads with optional parameters
+        {
+            if (stringBuilder is null)
+                throw new ArgumentNullException(nameof(stringBuilder));
+
+            // If the resulting string would end up on the large object heap, then use LargeText.
+            if (stringBuilder.Length >= LargeObjectHeapLimitInChars)
+            {
+                using var reader = new StringBuilderReader(stringBuilder);
```

Comment:
It seems like using a `TextReader` over the `StringBuilder` and passing it to the existing APIs that take a `TextReader` isn't ideal, because `TextReader` is a type that supports a lot of operations and if the existing APIs are someday rewritten to use operations that are not overriden in `StringBuilderReader` (like `ReadLine`), and therefore would use the base implementations in `TextReader`, it could silently tank the perf of this scenario. That's why I've overriden as many methods as made sense to me. But I guess it's better than duplicating the `LargeText.Decode` method and having duplicate code there.